swap/src/bitcoin/wallet.rs

```diff
@@ -202,106 +326,79 @@ impl Wallet {
-
-        let height = retry(ConstantBackoff::new(Duration::from_secs(1)), || async {
-            let height = reqwest::get(url.clone())
-                .await
-                .map_err(Error::Io)?
-                .text()
-                .await
-                .map_err(Error::Io)?
-                .parse::<u32>()
-                .map_err(|err| backoff::Error::Permanent(Error::Parse(err)))?;
-            Result::<_, backoff::Error<Error>>::Ok(height)
-        })
-        .await
-        .context("Transient errors should be retried")?;
+        let mut inner = self.client.lock().await;
 
-        Ok(BlockHeight::new(height))
+        inner.drain_notifications()?;
+
+        Ok(inner.latest_block)
     }
 
-    pub async fn transaction_block_height(&self, txid: Txid) -> Result<BlockHeight> {
+    pub async fn transaction_block_height(&self, txid: Txid) -> Result<Option<BlockHeight>> {
         let status_url = make_tx_status_url(&self.http_url, txid)?;
 
         #[derive(Serialize, Deserialize, Debug, Clone)]
         struct TransactionStatus {
             block_height: Option<u32>,
             confirmed: bool,
         }
-        let height = retry(ConstantBackoff::new(Duration::from_secs(1)), || async {
-            let block_height = reqwest::get(status_url.clone())
-                .await
-                .map_err(|err| backoff::Error::Transient(Error::Io(err)))?
-                .json::<TransactionStatus>()
-                .await
-                .map_err(|err| backoff::Error::Permanent(Error::JsonDeserialization(err)))?
-                .block_height
-                .ok_or(backoff::Error::Transient(Error::NotYetMined))?;
-
-            Result::<_, backoff::Error<Error>>::Ok(block_height)
-        })
-        .await
-        .context("Transient errors should be retried")?;
 
-        Ok(BlockHeight::new(height))
+        let block_height = reqwest::get(status_url.clone())
+            .await
+            .context("Failed to send request")?
+            .json::<TransactionStatus>()
+            .await
+            .context("Failed to deserialize response as TransactionStatus")?
+            .block_height;
+
+        Ok(block_height.map(BlockHeight::new))
     }
 
     pub async fn wait_for_transaction_finality(
         &self,
         txid: Txid,
+        script_to_watch: Script,
         execution_params: ExecutionParams,
     ) -> Result<()> {
         let conf_target = execution_params.bitcoin_finality_confirmations;
 
         tracing::info!(%txid, "Waiting for {} confirmation{} of Bitcoin transaction", conf_target, if conf_target > 1 { "s" } else { "" });
 
-        // Divide by 4 to not check too often yet still be aware of the new block early
-        // on.
-        let mut interval = interval(execution_params.bitcoin_avg_block_time / 4);
+        self.watch_until_status(txid, script_to_watch, |status| {
+            // TODO: Log the number of confirmations in here to the user.
```

Comment:
That would be good, need to make sure that we don't repeat it unnecessarily, i.e. only print `1 / 10`, `2 / 10`, etc

Make it follow the format that we have on the Monero side.